core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java

```diff
@@ -117,19 +122,45 @@ final class HttpSessionHandler extends ChannelDuplexHandler implements HttpSessi
     @Nullable
     private SessionProtocol retryProtocol;
 
+    /**
+     * {@code true} if an {@link Http2Settings} has been received from the remote endpoint.
+     */
+    private boolean isSettingsFrameReceived;
+    // Note: This field never becomes false once it becomes true.
+    private boolean channelActivated;
+
     HttpSessionHandler(HttpChannelPool channelPool, Channel channel,
-                       Promise<Channel> sessionPromise, ScheduledFuture<?> sessionTimeoutFuture,
+                       Promise<Channel> sessionPromise, int connectionTimeoutMillis,
                        SessionProtocol desiredProtocol, SerializationFormat serializationFormat,
                        PoolKey poolKey, HttpClientFactory clientFactory) {
         this.channelPool = requireNonNull(channelPool, "channelPool");
         this.channel = requireNonNull(channel, "channel");
         remoteAddress = channel.remoteAddress();
         this.sessionPromise = requireNonNull(sessionPromise, "sessionPromise");
-        this.sessionTimeoutFuture = requireNonNull(sessionTimeoutFuture, "sessionTimeoutFuture");
+        this.connectionTimeoutMillis = connectionTimeoutMillis;
         this.desiredProtocol = desiredProtocol;
         this.serializationFormat = serializationFormat;
         this.poolKey = poolKey;
         this.clientFactory = clientFactory;
+
+        if (poolKey.proxyConfig == ProxyConfig.direct()) {
+            scheduleSessionTimeout(channel, sessionPromise, connectionTimeoutMillis, desiredProtocol);
+        } else {
+            // A session timeout for a proxied connection may be scheduled when ProxyConnectionEvent is
```

Comment:
Question) I just wanted to make sure that even if we don't specify a connection timeout for `HaProxyHandler` the `connectTimeout` is respected.

i.e.
```
if (!poolKey.proxyConfig.proxyType().isForwardProxy())
```

Reply:
Good question. I didn't know that `HaProxyHandler` is not a subclass of `ProxyHandler`.  I made an additional PR to address your comment. #5268